Modularise DoLookupProtocol

vishalmohanty opened the pull request:
This diff does the following

- Abstracts out the `DoLookupProtocol` from `alookup` so that it can be used by other modules.
- Add UTs for `DoLookupProtocol` by mocking the `DoMiekgLookup` using Interface.

vishalmohanty committed:
Modularise DoLookupProtocol

spencerdrak commented on pkg/miekg/miekg_test.go:
```diff
@@ -1,12 +1,327 @@
+	if res, ok := mockResults[q.Name]; ok {
+		return res, nil, status, nil
+	} else {
+		return nil, nil, zdns.STATUS_NXDOMAIN, nil
```
Are there any other statuses that we may want to return here and see how the code handles it? STATUS_NXDOMAIN seems like it'd be the most common, but maybe in some corner cases we'd get something else.

vishalmohanty replied:
We have the `status` global variable for that. Look at `TestServFail`

zakird commented on pkg/miekg/miekg.go:
```diff
@@ -650,6 +663,93 @@ func (s *Lookup) DoMiekgLookup(q Question, nameServer string) (interface{}, zdns
+	}
+}
+
+func (s *Lookup) DoProtocolLookup(lc LookupClient, name string, nameServer string, dnsType uint16, candidateSet map[string][]Answer, cnameSet map[string][]Answer, origName string, depth int) ([]string, []interface{}, zdns.Status, error) {
```
If I'm reading this code correctly, this function does A/AAAA lookups with CNAME following, correct? I suspect that we can come up with a better function name that makes this clearer. The current name is very generic.

vishalmohanty replied:
Yes. I changed it to `DoIpsLookup`

spencerdrak commented on pkg/miekg/miekg.go:
```diff
@@ -1,3 +1,6 @@
+//go:build !mock
```
What's the uses for the build constraints here? Do we need to pass any flags to build mocks while testing?

spencerdrak replied:
rather, does this need to be excluded from any builds during testing?

vishalmohanty replied:
Sorry, this was from another thing I was trying. Removed now

spencerdrak commented on pkg/miekg/miekg.go:
```diff
@@ -74,6 +77,16 @@ type GlobalLookupFactory struct {
+}
+
+// Check whether the status is safe
+func SafeStatus(status zdns.Status) bool {
```
If this is used anywhere else (I'm not sure if the modules will have any use for this, or the CLI) it may be better defined in the util package.

vishalmohanty replied:
Moved to util

spencerdrak reviewed (commented):
I'd be curious if anyone else has feedback on how we've chosen to structure the mocking capability though. I think it's sensible, but another set of eyes would be good.

vishalmohanty committed:
Modularise DoTargetedLookup

vishalmohanty commented:
Summary of changes

- Moved `DoTargetedLookup` out of `alookup` as that can be utilised by other modules as well.
- Moved the UTs from `alookup` to `miekg` as all the logic now resides in `miekg`.
- Added basic UTs in `alookup` to ensure the flags are read correctly.

vishalmohanty committed:
Move VerifyAddress to util

spencerdrak reviewed (approved):
@vishalmohanty this looks good to me. We may have some additional work to expand tests further in the future, but this looks good to me!

gakiwate commented on pkg/miekg/miekg_test.go:
```diff
@@ -1,13 +1,483 @@
+	verifyResult(t, res.(IpResult), []string{"192.0.2.1", "192.0.2.2"}, nil)
+}
+
+func TestQuadAWithoutFlag(t *testing.T) {
```
What is this function trying to test? 

Also, generally could we add a line to each of the test cases as to the goal of the UT?

vishalmohanty replied:
Added comments

gakiwate commented on pkg/miekg/miekg_test.go:
```diff
@@ -1,13 +1,483 @@
+	verifyResult(t, res.(IpResult), nil, []string{"2001:db8::1", "2001:db8::2"})
+}
+
+func TestNoResults(t *testing.T) {
```
What does it mean for there to be no answers and the status to be NOERROR? Is this the empty non terminal case?

vishalmohanty replied:
Updated to return empty result in case of `NOERROR`

gakiwate commented on pkg/miekg/miekg_test.go:
```diff
@@ -1,13 +1,483 @@
+	verifyResult(t, res.(IpResult), nil, []string{"2001:db8::3"})
+}
+
+func TestUnexpectedMxOnly(t *testing.T) {
```
What is being tested here?

vishalmohanty replied:
Added comment

vishalmohanty committed:
Add comments to tests and return empty result with NOERROR
